### PR TITLE
[TE]frontend - Build the tree parser for composite anomalies

### PR DESCRIPTION
**Pull request by tejasajmera**
The parser contains the logic to extract data in order to build child components at any level of the tree when performing the drill-down analysis for Entity Monitoring. The output at each level of drill-down contains data formatted in a way that can be directly consumed by the child components referenced in the output without any additional transformation.

### CODE REVIEW

File: thirdeye/thirdeye-frontend/app/utils/anomalies-tree-parser.js [hunk 1]
```diff
@@ -0,0 +1,563 @@
+import { isEmpty } from "@ember/utils";
+import { set } from "@ember/object";
+import moment from "moment";
+
+const CLASSIFICATIONS = {
+  METRICS: {
+    KEY: "metrics",
```
Comment (zhangloo333) [inline thirdeye/thirdeye-frontend/app/utils/anomalies-tree-parser.js:7]: please keep code style consistency. I remember we enable the single quote rule. You could run prettier to format it.

File: thirdeye/thirdeye-frontend/app/utils/anomalies-tree-parser.js [hunk 2]
```diff
@@ -0,0 +1,563 @@
...
+    COMPONENT_PATH: "entity-groups",
+    DEFAULT_TITLE: "ENTITY:"
+  },
+  ENTITIES: {
+    KEY: "entities",
+    COMPONENT_PATH: "parent-anomalies",
+    DEFAULT_TITLE: "Entity"
+  }
+};
+const BREADCRUMB_TIME_DISPLAY_FORMAT = "MMM D HH:mm";
```
Comment (zhangloo333) [inline thirdeye/thirdeye-frontend/app/utils/anomalies-tree-parser.js:22]: I suggest you put this time format in the constant.js or to create a constant folder. it's a common useful variable. We can inject into the function.

File: thirdeye/thirdeye-frontend/app/utils/anomalies-tree-parser.js [hunk 3]
```diff
@@ -0,0 +1,563 @@
...
+ * Parse the anomalies generated by the composite alert to populate parent-anomalies table with relevent details about
+ * children for each anomaly.
+ *
+ * @param {Array<Object>} input
+ *   The anomalies for composite alert.
+ *
+ * @returns {Array<Object>}
+ *   Parsed out contents to populate parent-anomalies table
+ */
+const populateParentAnomaliesTable = input => {
```
Comment (zhangloo333) [inline thirdeye/thirdeye-frontend/app/utils/anomalies-tree-parser.js:52]: Please add parentheses for input to keep clarity and obey the eslint rule(arrow-parens).
For more detail to check `8.4 Always include parentheses around arguments for clarity and consistency. eslint: arrow-parens`

File: thirdeye/thirdeye-frontend/app/utils/anomalies-tree-parser.js [hunk 4]
```diff
@@ -0,0 +1,563 @@
...
+      feedback
+    };
+
+    const details = {};
+    let item;
+    if (children.length > 0) {
+      for (const child of children) {
+        const { metric, properties: { subEntityName } = {} } = child;
+
+        if (!isEmpty(metric)) {
```
Comment (zhangloo333) [inline thirdeye/thirdeye-frontend/app/utils/anomalies-tree-parser.js:70]: if there is no complicated logic, you could use the ternary to shorter logic and reduce code.

File: thirdeye/thirdeye-frontend/app/utils/anomalies-tree-parser.js [hunk 5]
```diff
@@ -0,0 +1,563 @@
...
+ * Parse the generated bucket for metric anomalies into the schema for the entity-metrics component
+ *
+ * @param {Object} input
+ *   The metric anomalies bucket constituents
+ *
+ * @returns {Array<Object>}
+ *   The content to be passed into the the leaf level entity-metrics component. Each item in the array represents
+ *   contents for the row in the table.
+ */
+const parseMetricsBucket = input => {
```
Comment (zhangloo333) [inline thirdeye/thirdeye-frontend/app/utils/anomalies-tree-parser.js:100]: ditto, coding style

File: thirdeye/thirdeye-frontend/app/utils/anomalies-tree-parser.js [hunk 6]
```diff
@@ -0,0 +1,563 @@
...
+ *
+ * @return {Object}
+ *   The anomaly referenced by id
+ */
+const findAnomaly = (id, input) => {
+  const { id: anomalyId, children } = input;
+
+  if (anomalyId === id) {
+    return input;
+  } else {
```
Comment (zhangloo333) [inline thirdeye/thirdeye-frontend/app/utils/anomalies-tree-parser.js:485]: Please keep the code concise. If an if block always executes a return statement, the subsequent else block is unnecessary. 
Please check the Airbnb rule. 16.3 or no-else-return

**Review by zhangloo333 — CHANGES_REQUESTED** [review]

File: thirdeye/thirdeye-frontend/app/utils/anomalies-tree-parser.js (hunk 1 above)
Comment (tejasajmera) [reply thirdeye/thirdeye-frontend/app/utils/anomalies-tree-parser.js:7]: Can you tell me what is the exact command for it? Is it tied to your code changes in https://github.com/apache/incubator-pinot/pull/6245? If so, I may not be able to do as your changes are not merged yet.

File: thirdeye/thirdeye-frontend/app/utils/anomalies-tree-parser.js (hunk 2 above)
Comment (tejasajmera) [reply thirdeye/thirdeye-frontend/app/utils/anomalies-tree-parser.js:22]: Sure.

File: thirdeye/thirdeye-frontend/app/utils/anomalies-tree-parser.js (hunk 3 above)
Comment (tejasajmera) [reply thirdeye/thirdeye-frontend/app/utils/anomalies-tree-parser.js:52]: Sure.

File: thirdeye/thirdeye-frontend/app/utils/anomalies-tree-parser.js (hunk 5 above)
Comment (tejasajmera) [reply thirdeye/thirdeye-frontend/app/utils/anomalies-tree-parser.js:100]: Sure.

File: thirdeye/thirdeye-frontend/app/utils/anomalies-tree-parser.js (hunk 4 above)
Comment (tejasajmera) [reply thirdeye/thirdeye-frontend/app/utils/anomalies-tree-parser.js:70]: Will do.

File: thirdeye/thirdeye-frontend/app/utils/anomalies-tree-parser.js (hunk 6 above)
Comment (tejasajmera) [reply thirdeye/thirdeye-frontend/app/utils/anomalies-tree-parser.js:485]: Yeah more than aiming for conciseness I was just trying to keeping it explicit. Anyway will remove the else.

File: thirdeye/thirdeye-frontend/app/utils/anomalies-tree-parser.js (hunk 1 above)
Comment (zhangloo333) [reply thirdeye/thirdeye-frontend/app/utils/anomalies-tree-parser.js:7]: I haven't realized my code hasn't merged yet. The command `npm run eslint-app` only can detect errors, but not fix them. You need to fix manually or set your vscode prettier with the rule of a single quote.

**Comment by tejasajmera** [comment]
@zhangloo333 Can you please take a look again, your comments are addressed now.

**Review by zhangloo333 — APPROVED** [review]
LGTM

**Review by harleyjj — APPROVED** [review]
LGTM

**Review by jihaozh — APPROVED** [review]
LGTM

**Pull request merged**